File: src/server/config/def/PocketMineServerProperties.php
<?php

namespace pocketcloud\cloud\server\config\def;

use pocketcloud\cloud\config\Config;
use pocketcloud\cloud\config\impl\MainConfig;
use pocketcloud\cloud\config\type\ConfigTypeList;
use pocketcloud\cloud\network\client\ServerClientCache;
use pocketcloud\cloud\network\Network;
use pocketcloud\cloud\server\CloudServer;
use pocketcloud\cloud\server\config\ServerProperties;
use pocketcloud\cloud\template\TemplateType;
use pocketcloud\cloud\util\Utils;
use const pocketcloud\CLOUD_PATH;

final class PocketMineServerProperties implements ServerProperties {

    public function modify(string $filePath, array $updatedContent): bool {
        $config = new Config($filePath, ConfigTypeList::PROPERTIES());
        foreach ($updatedContent as $name => $value) {
            $config->set($name, $value);
        }

        return $config->save();
    }

    public function renew(string $filePath): bool {
        $config = new Config($filePath, ConfigTypeList::PROPERTIES());
        $content = $config->getAll();
        Utils::fillMissingKeys($content, $this->getDefaultContent());
        $config->setAll($content);
        return $config->save();
    }

    public function needsRenewal(string $filePath): bool {
        if (file_exists($filePath)) {
            $config = new Config($filePath, ConfigTypeList::PROPERTIES());
            $keys = array_keys($this->getDefaultContent());
            $currentKeys = $config->getAll(true);
            return array_any($keys, fn(string $key) => !in_array($key, $currentKeys));
        }

        return true;
    }

    public function replacePlaceholders(CloudServer $server): array {
        return [
            "%uuid%" => $server->getServerUuid(),
            "%name%" => $server->getName(),
            "%server_port%" => $server->getServerData()->getPort(),
            "%server_portv6%" => $server->getServerData()->getPort() + 1,
            "%max_players%" => $server->getTemplate()->getMaxPlayerCount(),
            "%template%" => $server->getTemplate()->getName(),
            "%address%" => Network::getInstance()->getAddress()->getAddress(),
            "%port%" => Network::getInstance()->getAddress()->getPort(),
            "%encryption%" => MainConfig::getInstance()->isNetworkEncryptionEnabled(),
            "%language%" => "eng",
            "%cloud_path%" => CLOUD_PATH,
            "%timeout%" => $server->getTemplate()->getTemplateType()->getServerTimeout(),
            "%auth_key%" => Network::getInstance()->getAuthenticationKey(),
            "%server_ip%" => count(ServerClientCache::getInstance()->getAll(...TemplateType::onlyProxy())) > 0 ? "127.0.0.1" : "0.0.0.0"
        ];
    }

    public function getDefaultContent(): array {
        return [
            "language" => "eng",
            "motd" => "§b%name%",
            "server-port" => "%server_port%",
            "server-portv6" => "%server_portv6%",
            "server-ip" => "%server_ip%",
            "server-ipv6" => "::1",
            "enable-ipv6" => "on",
            "white-list" => "off",
            "max-players" => "%max_players%",
            "gamemode" => "SURVIVAL",
            "force-gamemode" => "off",
            "hardcore" => "off",
            "pvp" => "on",
            "difficulty" => 2,
            "generator-settings" => "",
            "level-name" => "world",
            "level-seed" => "",
            "level-type" => "DEFAULT",
            "enable-query" => "on",
            "auto-save" => "off",
            "view-distance" => 16,
            "xbox-auth" => "off",
            "server-uuid" => "%uuid%",
            "server-name" => "%name%",
            "template" => "%template%",
            "cloud-address" => "%address%",
            "cloud-port" => "%port%",
            "network-encryption" => "%encryption%",
            "cloud-language" => "%language%",
            "cloud-path" => "%cloud_path%",
            "server-timeout" => "%timeout%",
            "auth-key" => "%auth_key%"
        ];
    }

    public function getFileName(): string {
        return "server.properties";
    }

    public function getTemplateType(): TemplateType {
        return TemplateType::SERVER();
    }
}

File: src/main/java/de/pocketcloud/cloud/server/config/impl/PocketMineServerProperties.java
package de.pocketcloud.cloud.server.config.impl;

import de.pocketcloud.cloud.PocketCloud;
import de.pocketcloud.cloud.config.impl.MainConfig;
import de.pocketcloud.cloud.network.client.ServerClientCache;
import de.pocketcloud.cloud.server.CloudServer;
import de.pocketcloud.cloud.server.config.ServerProperties;
import de.pocketcloud.cloud.server.software.ServerSoftware;
import de.pocketcloud.cloud.server.software.ServerSoftwareManager;
import de.pocketcloud.cloud.util.ArrayUtils;

import java.util.LinkedHashMap;
import java.util.Map;

public final class PocketMineServerProperties extends ServerProperties {

    @Override
    public boolean modify(String filePath, Map<String, Object> updatedContent) {
        return modifyProperties(filePath, updatedContent);
    }

    @Override
    public boolean renew(String filePath) {
        return renewProperties(filePath);
    }

    @Override
    public boolean needsRenewal(String filePath) {
        return needsRenewalProperties(filePath);
    }

    @Override
    public Map<String, Object> replacePlaceholders(CloudServer server) {
        return new LinkedHashMap<>(Map.ofEntries(
            Map.entry("%uuid%", server.uuid().toString()),
            Map.entry("%name%", server.name()),
            Map.entry("%server_port%", server.serverData().port()),
            Map.entry("%server_portv6%", server.serverData().port() + 1),
            Map.entry("%max_players%", server.template().settings().getMaxPlayerCount()),
            Map.entry("%template%", server.templateName()),
            Map.entry("%address%", MainConfig.instance().getNetworkAddress().getHostString()),
            Map.entry("%port%", MainConfig.instance().getNetworkAddress().getPort()),
            Map.entry("%encryption%", MainConfig.instance().isNetworkEncryptionEnabled()),
            Map.entry("%language%", MainConfig.instance().getLanguage()),
            Map.entry("%cloud_path%", System.getProperty("user.dir")),
            Map.entry("%timeout%", server.template().templateType().timeout()),
            Map.entry("%auth_key%", PocketCloud.instance().network().authToken()),
            Map.entry("%server_ip%", !ServerClientCache.instance().getAll(c -> {
                if (c.hasServer()) {
                    return c.server().template().templateType().isProxy();
                }

                return false;
            }).isEmpty() ? "127.0.0.1" : "0.0.0.0"),
            Map.entry("%packet_size_limit%", MainConfig.instance().getNetworkPacketSizeLimit())
        ));
    }

    @Override
    public Map<String, Object> getDefaultContent() {
        return ArrayUtils.orderedMap(
                "language", "eng",
                "motd", "§b%name%",
                "server-port", "%server_port%",
                "server-portv6", "%server_portv6%",
                "server-ip", "%server_ip%",
                "server-ipv6", "::1",
                "enable-ipv6", "on",
                "white-list", "off",
                "max-players", "%max_players%",
                "gamemode", "SURVIVAL",
                "force-gamemode", "off",
                "hardcore", "off",
                "pvp", "on",
                "difficulty", 2,
                "generator-settings", "",
                "level-name", "world",
                "level-seed", "",
                "level-type", "DEFAULT",
                "enable-query", "on",
                "auto-save", "off",
                "view-distance", 16,
                "xbox-auth", "off",
                "server-uuid", "%uuid%",
                "server-name", "%name%",
                "template", "%template%",
                "cloud-address", "%address%",
                "cloud-port", "%port%",
                "network-encryption", "%encryption%",
                "cloud-language", "%language%",
                "cloud-path", "%cloud_path%",
                "server-timeout", "%timeout%",
                "auth-key", "%auth_key%",
                "packet-size-limit", "%packet_size_limit%"
        );
    }

    @Override
    public String getFileName() {
        return "server.properties";
    }

    @Override
    public ServerSoftware getServerSoftware() {
        return ServerSoftwareManager.instance().get("pmmp-latest");
    }
}